### ISSUE TITLE
name resolution isn't namespaced

### ISSUE THREAD

**Issue by npmccallum** [WebAssembly/wasi-sockets#15]
# Problem

A `tcp-socket` is [created with ](https://github.com/WebAssembly/wasi-sockets/blob/main/wasi-socket-tcp.wit#L15)a `network` reference. OTOH, [resolving an address](https://github.com/WebAssembly/wasi-sockets/blob/main/wasi-ip-name-lookup.wit#L32) has no `network` context. Likewise, [IP addresses](https://github.com/WebAssembly/wasi-sockets/blob/main/common-types.wit#L19-L32) have no `network` context.

This means that resolving a name could return an address that is valid on both `network`s but points to different hosts.

# Example

Resolving `example.com` returns `10.1.2.3`. There are two `network`s present. The first has an interface on `10.0.0.0/8` and the second has an interface on a different physical network with `10.1.0.0/16`. `10.1.2.3` is valid on both `network`s; but is a different physical system. On which `network` should I create a `tcp-socket` in order to reach `example.com`?

# Resolution

Name resolution cannot be global if sockets are namespaced. Either name resolution must be namespaced or socket creation must not be.

**Comment by badeend** [comment]
Thanks for noticing! This was definitely an oversight. `resolve-addresses` should take a `network` handle.

Just to clarify, the `network` handle represents the ability to use networking operations. It does not (necessarily) represent a physical single network interface. The definition of "network" is intentionally left vague, to let each implementation decide for themselves how they wish to slice and dice network access.

**Comment by npmccallum** [comment]
@badeend I think you've missed my point. Not only does `resolve-addresses` need to take a `network` handle but if you choose to go that way you can **never** refer to an ip address by itself. This cuts across the grain of literally every networking library ever created. Therefore, following this route means that you have opted for maximum incompatibility with all existing code.

I think the more sensible route is dropping the `network` context altogether. This does cut against the grain of context-based-security. However, this matches the behavior of every operating system and code in existence.

**Comment by badeend** [comment]
> I think you've missed my point. Not only does resolve-addresses need to take a network handle but if you choose to go that way you can never refer to an ip address by itself. 

Could you lead me through your thought process on how you arrive at that conclusion?

> I think the more sensible route is dropping the network context altogether. This does cut against the grain of context-based-security. However, this matches the behavior of every operating system and code in existence.

That would indeed go against the [WASI Design Principles](https://github.com/WebAssembly/wasi#wasi-design-principles).

**Comment by npmccallum** [comment]
> Could you lead me through your thought process on how you arrive at that conclusion?

If there is only one network "view" for a wasm instance (i.e. one global `network`), then all IP addresses are unambiguous. This is the way existing operating systems work. For example, on Linux although you can have multiple network namespaces there can only be one current namespace. Therefore, there is a single network "view" and all IP addresses within that view are unambiguous.

However, if all network operations are now contextualized on a `network` handle, then a single ip address is now an ambiguous reference to a remote entity. To make an ip address unambiguous it also needs a `network` handle. Therefore, an unambiguous remote reference is `tuple<network, ip-address>`.

Talking to the wrong remote entity can definitely result in data disclosure and is therefore a security problem.

**Comment by badeend** [comment]
Doesn't the same issue apply to directory file descriptors (networks) and string file paths (ipaddresses)?

```rs
write_at(dir_fd, "password.txt", "<<very sensitive data>>")
```

Here, the path "password.txt" alone is ambiguous too. If `dir_fd` is the user's home directory it might be fine to write the data. If `dir_fd` points to the `public_html` folder of a webserver, the same code suddenly isn't.

---

> If there is only one network (...), then all IP addresses are unambiguous.

IP-addresses are always contextual, even when there is only one `network` handle (or: none). Example: `10.0.0.0` probably resolves to something different on my machine than on yours.


It feels like you're trying to impose some kind of meaning onto bare IP-addresses that the network stack can't guarantee. (at least not at the posix-socket level of abstraction)

Let me know what you think.

**Comment by npmccallum** [comment]
@badeend 

> Here, the path "password.txt" alone is ambiguous too.

The word **alone** is key here. File base names are always contextual. And APIs are written to handle that context.

> IP-addresses are always contextual

This statement is false. IP addresses are globally unique UNLESS a particular range is defined. Your example of `10.0.0.0` only works because `10.0.0.0/8` is a reserved range as defined by [RFC 1918](https://datatracker.ietf.org/doc/html/rfc1918#section-3). All addresses are globally unique except the reserved ranges.

**Comment by sunfishcode** [comment]
@npmccallum 

> > IP-addresses are always contextual

> This statement is false. IP addresses are globally unique UNLESS a particular range is defined. Your example of 10.0.0.0 only works because 10.0.0.0/8 is a reserved range as defined by [RFC 1918](https://datatracker.ietf.org/doc/html/rfc1918#section-3). All addresses are globally unique except the reserved ranges.

IP addresses may be globally unique on the public internet, but one of the goals of WASI is to make APIs virtualizable. A WASI program shouldn't be able to tell if it's talking to the public internet or not, because sometimes it's useful to fully virtualize a network, and that includes virtualizing how a non-reserved-range address like 198.35.26.96 is resolved.

**Comment by npmccallum** [comment]
@sunfishcode I can accept that.

I don't see any way around the broader "interface" problem without providing a "default network" however.

**Comment by badeend** [comment]
@npmccallum This issue hasn't seen recent activity and, if I understand the discussion above correctly, there are no outstanding actionable items. Do you mind if I close this?

FYI, `resolve-addresses` now takes a `network` parameter. (Since https://github.com/WebAssembly/wasi-sockets/pull/16)

**Comment by sunfishcode** [comment]
The issue here is resolved in #16, with `resolve-addresses` now taking a `network` parameter, so I think it's ok to close this. If there are any further concerns, please file new issues!
